lighthouse-core/audits/dobetterweb/uses-responsive-images.js

```diff
@@ -0,0 +1,139 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+ /**
+  * @fileoverview Checks to see if the images used on the page are larger than
+  *   their display sizes. The audit will list all images that are larger than
+  *   their display size regardless of DPR (a 1000px wide image displayed as a
+  *   500px high-res image on a Retina display will show up as 75% unused);
+  *   however, the audit will only fail pages that use images that have waste
+  *   when computed with DPR taken into account.
+  */
+'use strict';
+
+const Audit = require('../audit');
+const URL = require('../../lib/url-shim');
+const Formatter = require('../../formatters/formatter');
+
+const KB_IN_BYTES = 1024;
+const WASTEFUL_THRESHOLD_AS_RATIO = 0.1;
+
+class UsesResponsiveImages extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Images',
+      name: 'uses-responsive-images',
+      description: 'Site uses appropriate image sizes',
+      helpText: 'Image sizes served should be based on the display size to save network bytes. ' +
+        'Learn more about [responsive images](https://developers.google.com/web/fundamentals/design-and-ui/media/images) ' +
+        'and [client hints](https://developers.google.com/web/updates/2015/09/automating-resource-selection-with-client-hints).',
+      requiredArtifacts: ['ImageUsage', 'ContentWidth']
+    };
+  }
+
+  /**
+   * @param {!Object} image
+   * @param {number} DPR
+   * @return {?Object}
+   */
+  static computeWaste(image, DPR) {
+    const url = URL.getDisplayName(image.src);
+    const actualPixels = image.naturalWidth * image.naturalHeight;
+    const usedPixels = image.clientWidth * image.clientHeight;
+    const usedPixelsFullDPR = usedPixels * Math.pow(DPR, 2);
+    const usedRatio = usedPixels / actualPixels;
+    const usedRatioFullDPR = usedPixelsFullDPR / actualPixels;
+
+    if (!usedRatio || usedRatio >= 1) {
```

Comment:
Yeah will better handle the edge cases, originally this was for an image that was never shown (0 used pixels) but that is now handled by checking for the network record